src/config.js

```diff
@@ -0,0 +1,61 @@
+/* @flow */
+import path from 'path';
+
+import requireUncached from 'require-uncached';
+
+import {createLogger} from './util/logger';
+import {UsageError} from './errors';
+
+const log = createLogger(__filename);
+
+type ApplyConfigToArgvParams = {
+  argv: Object,
+  configObject?: Object,
+  configFileName?: string,
+  defaultValues?: Object,
+};
+
+type ParseConfigParams = {
+  argv: Object,
+  configFileName: string,
+}
+
+export function applyConfigToArgv({
+  argv,
+  configObject,
+  configFileName,
+  defaultValues = {},
+}: ApplyConfigToArgvParams): Object {
+  if (configFileName) {
+    configObject = parseConfig(argv, configFileName);
+  }
+  for (const option in configObject) {
+    if (!argv.hasOwnProperty(option) || defaultValues[option]) {
+      argv[option] = configObject[option];
+    }
+  }
+  return argv;
+}
+
+export function parseConfig({
+  argv,
```

Comment:
This small change is making ~25 completely unrelated tests fail. I've addressed all the other nits and feedback, will take some time to get to the bottom of this one.